whylog/config/rule.py

```diff
@@ -19,10 +18,10 @@ def serialize(self):
             "constraints": self._constraints,
         }
 
-    def get_new_parsers(self, old_parsers):
+    def get_new_parsers(self, parser_name_generator):
         new_parsers = []
-        for parser in itertools.chain([self._effect], self._causes):
-            if parser.name not in old_parsers:
+        for parser in self._causes + [self._effect]:
+            if parser_name_generator.is_free_parser_name(parser.name):
```

Comment:
Here may occurs problem with mulithreading teachers. In single thred teachers is no problem